File: src/endpoints/app/put_item.py
from . import __types__ as types
from src import output
from src.libs.db import DB, Table
from src.modules.verify import checking_token

async def put_item(params: dict = {}, req = None, _db: DB = None, _token = None):

    # set values
    result = None
    db = _db if _db else DB().connect()

    try:
        # set params
        params = types.PutItem(**params)

        # checking token
        token = checking_token(req, db) if not _token else _token

        # check code already exists
        count = db.get_count(
            table_name = Table.APP.value,
            where = [ f'code LIKE \'{params.code}\'' ],
        )
        if count > 0: raise Exception('code already exists', 400)

        # set values
        values = {
            'code': params.code,
            'name': params.name,
            'description': params.description or '',
        }

        # set keys
        placeholders = [
            { 'key': 'code', 'value': ':code' },
            { 'key': 'name', 'value': ':name' },
            { 'key': 'description', 'value': ':description' },
            { 'key': 'created_at', 'value': 'DATETIME("now", "localtime")' },
        ]

        # add item
        data = db.add_item(
            table_name = Table.APP.value,
            placeholders = placeholders,
            values = values,
        )

        # set result
        result = output.success({
            'message': 'Complete add item.',
            'data': data,
        }, _req=req)
    except Exception as e:
        result = output.exc(e, _req=req)
    finally:
        if not _db and db: db.disconnect()
        return result

File: src/routes/app/put-item.ts
import DB, { db } from '@/classes/DB'
import ServiceError from '@/classes/ServiceError'
import * as messages from '@/libs/messages'
import { printf } from '@/libs/strings'
import type { AppModel } from './__model'

type PutItemParams = {
  body: AppModel['putItemBody']
}

export default async function putItem({ body }: PutItemParams)
{
  try
  {
    // check exist data
    const count = db.getCount({
      table: DB.TABLE.APP,
      where: 'code = $code',
      values: { '$code': body.code },
    })
    if ((count.data as number) > 0)
    {
      throw new ServiceError(printf(messages.ERR_EXISTS, 'code'), { status: 400 })
    }

    // add data
    const addedData = db.addData({
      table: DB.TABLE.APP,
      values: [
        { key: 'code', value: body.code },
        { key: 'name', value: body.name },
        body.description && { key: 'description', value: body.description },
        { key: 'created_at', valueName: DB.DATE_TIME },
      ].filter(Boolean),
    })

    return addedData.data || 0
  }
  catch (_e: any)
  {
    throw new ServiceError('Failed to add App.', {
      status: _e.status,
      text: _e.message,
      cause: _e,
    })
  }
}
